### DISCUSSION TITLE
JavaScript for Ajax in an Admin View

### DISCUSSION THREAD

**Discussion by linpi1** [joomengine/Joomla-Component-Builder#806] [Q&A] [answered]
in an admin view for customers, I would like to get the coupon value after I entering the coupon (+ return) with ajax. 
both php files ajax.php and ajax.json.php seams fine to me, just the javascript is not easy for me...  

I entered in the coupon field to get the value of the coupon after a return, an firing the controller...
`jQuery('#jform_author_coupon').change(function(){
getCoupon(jQuery('#jform_author_coupon').val());
});`

and in view footer
`function getCoupon(coupon){
	var getUrl = JRouter("index.php?option=com_bookshop&task=ajax.getCoupon&format=json&raw=true");
	if(token.length > 0 && user > 0){
		var request = 'token='+token+'&coupon='+coupon;
	}
	return jQuery.ajax({
		type: 'GET',
		url: getUrl,
		dataType: 'json',
		data: request,
		jsonp: false
	});
}`
the function getCoupon is called, but I get in the console, this error: "JRouter is not defined" 

I know, the ajax code is not even finished, since the result is not echoing in the coupon value field. 
It is a little embarrassing to ask the kind of questions, I'm just  blocked... and obviously I'm not a JavaScript master... ;-)  
thank you for a little hint.

**Comment by linpi1** [comment]
just to close to this post:
finally and temporarily it works, but I didn't found why JRouter is not defined.  
I hade to modify the ajax javascript:
`    function getCoupon(coupon){
        var getUrl = 'index.php?option=com_bookshop&task=ajax.getCoupon&format=json&raw=true';
    	if(token.length > 0){
    		var request = 'token='+token+'&coupon='+coupon;
            jQuery.ajax({
                		type: 'GET',
                		url: getUrl,
                		dataType: 'json',
                		data: request,
                		jsonp: false,
                        success: function(data) {
                		    jQuery('#jform_coupon_value_author_str').val(data);
                        }
                	});
    	}
    }`
probably not elegant but it works...

**Comment by TLWebdesign** [comment, marked as answer]
JRouter refers to a Javascript function that needs to be included in your view's footer (JS). It's a custom code llewellyn uses in his extensions (and i think most of us too) it should be there in "custom codes". If not try installing the demo package in JCB and it should appear in custom codes view.
<img width="1245" alt="Schermafbeelding 2021-09-28 om 17 57 25" src="https://user-images.githubusercontent.com/4402824/135123210-c8da90d6-ae3e-4805-9d03-fd09e3dcac37.png">

Add the "[CUSTOMCODE=jsRouter]" tag  to the Add JavaScript (view-footer) * field in your admin view under the "javascript" tab.

Then the code should work.

**Reply by linpi1** [reply]
Cool thank's a lot, just fantastic !
